File: client/lib/dispatch/subrequest.dart
/// Wire-shape value classes for one sub-request and one sub-response.
///
/// These mirror REQUIREMENTS.md §4.1 N-API-2/3. They are intentionally
/// untyped at the `data` level — the [Dispatcher] handles encode/decode of
/// the typed payloads via the handler registry.
library;

/// One element in the batched request body sent to `POST /api/v1/batch`.
class SubRequest {
  /// Client-supplied correlation id. The dispatcher generates a UUID for
  /// every call; tests may pass an explicit id for stable assertions.
  final String id;

  /// `data | action | query` per the requirements. Defaults to `data`.
  final String type;
  final String endpoint;
  final String action;

  final Map<String, dynamic> ref;
  final Map<String, dynamic> key;

  /// JSON-encodable payload. May be `null` for handlers like `card_type.select`
  /// that take no input.
  final dynamic data;

  const SubRequest({
    required this.id,
    required this.type,
    required this.endpoint,
    required this.action,
    this.ref = const {},
    this.key = const {},
    this.data,
  });

  Map<String, dynamic> toJson() {
    final m = <String, dynamic>{
      'id': id,
      'type': type,
      'endpoint': endpoint,
      'action': action,
    };
    if (ref.isNotEmpty) m['ref'] = ref;
    if (key.isNotEmpty) m['key'] = key;
    if (data != null) m['data'] = data;
    return m;
  }
}

/// One element in the batched response body. The decoder is intentionally
/// permissive with `data` — the dispatcher hands the raw JSON tree off to
/// the registered output codec.
class SubResponse {
  final String id;
  final bool ok;
  final dynamic data;
  final SubResponseError? error;

  const SubResponse({
    required this.id,
    required this.ok,
    this.data,
    this.error,
  });

  factory SubResponse.fromJson(Map<String, dynamic> json) {
    SubResponseError? err;
    final raw = json['error'];
    if (raw is Map<String, dynamic>) {
      err = SubResponseError(
        (raw['code'] as String?) ?? '',
        (raw['message'] as String?) ?? '',
      );
    }
    return SubResponse(
      id: json['id'] as String,
      ok: (json['ok'] as bool?) ?? false,
      data: json['data'],
      error: err,
    );
  }
}

class SubResponseError {
  final String code;
  final String message;
  const SubResponseError(this.code, this.message);
}

File: client/src/dispatch/subrequest.ts
/**
 * Wire-shape value types for one sub-request and one sub-response.
 *
 * Mirrors REQUIREMENTS.md §4.1 N-API-2/3. The `data` field is intentionally
 * untyped at this layer — the {@link Dispatcher} handles encode/decode of the
 * typed payloads via the handler registry.
 */

/** One element in the batched request body sent to `POST /api/v1/batch`. */
export interface SubRequest {
  /** Client-supplied correlation id (UUID v4). */
  id: string;
  /** `data | action | query` per the requirements. */
  type: string;
  endpoint: string;
  action: string;
  ref: Record<string, unknown>;
  key: Record<string, unknown>;
  /** JSON-encodable payload, or `null` for handlers that take no input. */
  data: unknown;
}

/** Per-subresponse error envelope. */
export interface SubError {
  code: string;
  message: string;
  /**
   * Optional structured payload the server carries over from
   * `reg.HandlerError.Detail`. Gate 5 (FLOW_AND_SCREEN_KERNEL §V13) uses
   * this for the positive-feedback rejection envelope on
   * `flow_disallowed` / `flow_role_required` — see
   * `client/src/ui/widgets/TransitionBar.svelte`. Other handlers may
   * leave it absent.
   */
  detail?: unknown;
}

/** One element in the batched response body. */
export interface SubResponse {
  id: string;
  ok: boolean;
  data?: unknown;
  error?: SubError;
}

/**
 * Encode a {@link SubRequest} for the wire. Mirrors the Dart `toJson` —
 * empty `ref`/`key` are omitted, and `data: null` is omitted.
 */
export function subRequestToJson(s: SubRequest): Record<string, unknown> {
  const m: Record<string, unknown> = {
    id: s.id,
    type: s.type,
    endpoint: s.endpoint,
    action: s.action,
  };
  if (Object.keys(s.ref).length > 0) m.ref = s.ref;
  if (Object.keys(s.key).length > 0) m.key = s.key;
  if (s.data !== null && s.data !== undefined) m.data = s.data;
  return m;
}

/** Permissive parser for a single sub-response from the server. */
export function subResponseFromJson(raw: Record<string, unknown>): SubResponse {
  let error: SubError | undefined;
  const e = raw['error'];
  if (e && typeof e === 'object') {
    const obj = e as Record<string, unknown>;
    error = {
      code: typeof obj['code'] === 'string' ? (obj['code'] as string) : '',
      message: typeof obj['message'] === 'string' ? (obj['message'] as string) : '',
    };
    if (obj['detail'] !== undefined && obj['detail'] !== null) {
      error.detail = obj['detail'];
    }
  }
  const out: SubResponse = {
    id: typeof raw['id'] === 'string' ? (raw['id'] as string) : '',
    ok: raw['ok'] === true,
    data: raw['data'],
  };
  if (error !== undefined) out.error = error;
  return out;
}
